gnucash/import-export/qif-imp/assistant-qif-import.c

```diff
@@ -1987,6 +1992,7 @@ gnc_ui_qif_import_load_progress_start_cb (GtkButton * button,
 
     gtk_widget_set_sensitive (wind->load_pause, FALSE);
     wind->busy = FALSE;
+    }
```

Comment:
The `if` condition this brace closes is `!wind->load_stop` and the following line is a new if with the same condition, so if this is really right then you should remove it too. But it's a major change: You're making an additional 120 lines execute only if `!wind->load_stop` that previously executed regardless of `load_stop`'s value. Are you sure that's right?

Aside from that the whole block needs to be indented.

Reply:
> The `if` condition this brace closes is `!wind->load_stop` and the following line is a new if with the same condition, so if this is really right then you should remove it too. But it's a major change: You're making an additional 120 lines execute only if `!wind->load_stop` that previously executed regardless of `load_stop`'s value. Are you sure that's right?

This is from testing with and without the `if` condition changes. I think 755c42a has created a flaw in the workflow and this commit fixes it.

> Aside from that the whole block needs to be indented.

Agreed, will be reindented properly iff the `if` change is confirmed correct.

Reply:
That's hand waving and you're not changing the condition, you're extending the block that it prevents from operating. What failed?

That aside, rather than extending the block you might as well just flip the test, set `wind->load_stop = FALSE;` and return.

Reply:
I find the C part confusing still. Maybe best I remove the C changes commit, expose the bug, and leave it for you?

Reply:
No surprise, it's a really ugly function with far too many moving parts. But I can move a brace around, just tell me what is the bug!

Reply:
Ok not at home atm to test; from memory if you apply this branch but revert 026ff3a0, and try load QIF with weird charsets available at https://bugs.gnucash.org/show_bug.cgi?id=797218, the incorrect-charset error isn't handled very well.

Reply:
i.e. without 026ff3a:
select shift-JIS.qif for import, chose iso-8859-1
incorrect-charset appears twice (from memory)
try press back - can't remember if it works.

the qif-import mechanism also allows loading multiple qif files (eg. from same quicken book, load each qif-account, to achieve mass import) and I'm not 100% sure how it should handle multiple-qif-files, each with its own charset

Reply:
With the braces as-is when I try to import shiftjis.qif using Unicode(UTF-8) the load completes. The message "The file contains characters that are not part of charset UTF-8. Please choose a different one." With the brace in its original position it also pops a message box saying "A bug was detected while parsing the QIF file." and repeats the same thing in the log window along with the first message. The back button works in either case.

shiftjis.qif imports with no error if you tell it Western(iso-8859-1), I suppose because there aren't any illegal bytes in iso-8859 charsets.

Incidentally, several of Phurrough's qif files fail on MacOS with a guile error "cannot convert narrow string to output locale" or "cannot convert wide string to output locale" from scm_to_stringn depending on what encoding one selects. Haven't tested on Windows yet.

Reply:
Haven't seen this narrow/wide char particular issue. Would this error come up only when an invalid charset is chosen? If so then guile's exception handler can catch it as well:

```patch
modified   gnucash/import-export/qif-imp/qif-file.scm
@@ -526,7 +526,7 @@
     (lambda (key . rest)
       (let ((msg (string-append
                   (case key
-                    ((decoding-error)
+                    ((decoding-error encoding-error)
                      (format #f (G_ "The file contains characters that are not \
 part of charset ~a. Please choose a different one.") encoding))
                     ((system-error) (G_ "I/O system error. Try again."))
```

Reply:
Well, this is disappointing: I set a breakpoint on the wide string to output locale error. Here's the bt:
```
(lldb) bt 16
* thread #1, queue = 'com.apple.main-thread', stop reason = breakpoint 2.1
  * frame #0: 0x00000001008d32ae libguile-2.2.1.dylib`scm_to_stringn(str=0x0000000151230b40, lenp=0x0000000000000000, encoding="ASCII", handler=SCM_FAILED_CONVERSION_ERROR) at strings.c:2270:39
    frame #1: 0x00000001008d2f91 libguile-2.2.1.dylib`scm_to_locale_stringn(str=0x0000000151230b40, lenp=0x0000000000000000) at strings.c:1909:10
    frame #2: 0x00000001008d2f49 libguile-2.2.1.dylib`scm_to_locale_string(str=0x0000000151230b40) at strings.c:1903:10
    frame #3: 0x0000000100917a17 libguile-2.2.1.dylib`scm_regexp_exec(rx=0x000000015128d4a0, str=0x0000000151230b40, start=0x0000000000000904, flags=0x0000000000000002) at regex-posix.c:274:11
    frame #4: 0x0000000100847901 libguile-2.2.1.dylib`scm_apply_subr(sp=0x000000011f4fbbf0, nslots=5) at gsubr.c:311:14
```

It's the same old UTF-8 conversion error passing through `scm_to_locale_string` problem that makes Guile unsuitable for string handling. You'll need to rewrite qif_file_load to C/C++.

Reply:
You know very well that I can't write C well... if @derekatkins could not write a qif importer in C then there's no chance I can. Any idea which guile function is calling `scm_to_locale_string`?

Reply:
Sorry, I thought it obvious. It's your attempt to pass `encoding` to `with-input-from-file` in `qif-file:read-file`. That only works on Linux because on MacOS the "locale" encoding is hard-coded to ASCII and on Windows to whatever is the active code page for CMD. You need to use C to read the contents of the file and transcode it to UTF8 and pass that to `scm-from-utf8-string`. *You ripped out the code that did that.*

Reply:
Ok. Ripping out the code was meant to allow guile to throw a `decoding-error` exception to report via UI. But "cannot convert narrow/wide string to output locale" seems to be a result of `encoding-error` in guile, hence, the idea to catch it as well *may* work?

Reply:
Sure, you can catch the exception. The problem is that it will get thrown *every time* you try to transcode on macOS or Windows because Guile's transcoding design using ports is irretrievably broken.

Reply:
Does the exception get thrown if the correct charset is selected?

Reply:
If there's a need for complete scm->C rewrite then I definitely won't be able to do it. If it's only 1-2 functions then maybe I can.

Reply:
> Does the exception get thrown if the correct charset is selected?

Yes.

>  If it's only 1-2 functions then maybe I can.

There are a couple of possible approaches. One would be to extend the existing C transcoding functions that `qif-file:read-file` already calls to take a proposed encoding argument and raise errors that Scheme can pass back through `qif-file:read-file`. The other would be to rewrite `qif-file:read-file` in C, returning an SCM string to pass on to the importer. I think the second would be cleaner because it gets guile out of the middle for transcoding, error handling, and UI.

Reply:
Ok -- this means handle the charset test and failure, and BOM stripping, all in C rather than SCM. It means the majority of `qif-file:read-file` to decode each line will remain in scheme. This much more reasonable.